test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc

```diff
@@ -0,0 +1,180 @@
+
+  timer_ = dispatcher_.createTimer([this]() -> void {
+    finished_ = true;
+    this->cb_.complete();
+    return;
+  });
+  timer_->enableTimer(timeout_0ms);
+}
+
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  (void)out;
+  (void)out_len;
+  (void)signature_algorithm;
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int hash_len;
+  EC_KEY* ec_key;
+  const EVP_MD* md;
+  bssl::ScopedEVP_MD_CTX ctx;
+  EcdsaPrivateKeyConnection* ops = static_cast<EcdsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index));
+
+  if (!ops) {
+    return ssl_private_key_failure;
+  }
+
+  ec_key = ops->getPrivateKey();
+  if (!ec_key) {
+    return ssl_private_key_failure;
+  }
+
+  md = SSL_get_signature_algorithm_digest(signature_algorithm);
+  if (!md) {
+    EC_KEY_free(ec_key);
+    return ssl_private_key_failure;
+  }
+
+  // Calculate the digest for signing.
+  if (!EVP_DigestInit_ex(ctx.get(), md, nullptr) || !EVP_DigestUpdate(ctx.get(), in, in_len) ||
+      !EVP_DigestFinal_ex(ctx.get(), hash, &hash_len)) {
+    EC_KEY_free(ec_key);
+    return ssl_private_key_failure;
+  }
+
+  ops->out_ = static_cast<uint8_t*>(OPENSSL_malloc(max_out));
+  if (ops->out_ == nullptr) {
+    EC_KEY_free(ec_key);
+    return ssl_private_key_failure;
+  }
+
+  if (!ECDSA_sign(0, hash, hash_len, ops->out_, &ops->out_len_, ec_key)) {
+    EC_KEY_free(ec_key);
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  EC_KEY_free(ec_key);
+  // Tell SSL socket that the operation is ready to be called again.
+  ops->delayed_op();
+
+  return ssl_private_key_retry;
+}
+
+static ssl_private_key_result_t privateKeyDecrypt(SSL* ssl, uint8_t* out, size_t* out_len,
+                                                  size_t max_out, const uint8_t* in,
+                                                  size_t in_len) {
+  (void)ssl;
+  (void)out;
+  (void)out_len;
+  (void)max_out;
+  (void)in;
+  (void)in_len;
+
+  return ssl_private_key_failure;
+}
+
+static ssl_private_key_result_t privateKeyComplete(SSL* ssl, uint8_t* out, size_t* out_len,
+                                                   size_t max_out) {
+  EcdsaPrivateKeyConnection* ops = static_cast<EcdsaPrivateKeyConnection*>(
+      SSL_get_ex_data(ssl, EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index));
+
+  if (!ops->finished_) {
+    // The operation didn't finish yet, retry.
+    return ssl_private_key_retry;
+  }
+
+  if (ops->test_options_.async_method_error_) {
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  if (ops->out_len_ > max_out) {
+    OPENSSL_free(ops->out_);
+    return ssl_private_key_failure;
+  }
+
+  memcpy(out, ops->out_, ops->out_len_);
```

Comment:
Should check that the amount of data copied is &le; `max_out`.

Reply:
Yes, that's right. Fixed.